### PR TITLE
fix: fix repl function application evaluation

### LINKED ISSUE #1237: Calling functions in the REPL is broken
```
鲤 (defn f [] 123)
鲤 (f)
I did not understand the form `((defn f [] 123))` at REPL:1:2.

Traceback:
  (f) at REPL:1:1.

```

### PR DESCRIPTION
**Pull request by scolsen**
The previous pattern matching changes introduced a subtle error in the
evaluator's handling of function applications, we used the same Resolver
for both lists and symbols, where previously we hadn't. Restoring the
old resolution selection and adding a new clause to capture forms like
((defn f [] 2)) restores the correct functioning in the repl.

Importantly, forms such as ((defn foo [x] x) 3), should only be resolved
when the defn form is the result of expanding a symbol.

Eventually, resolving defns to anonymous functions might simplify this
case.

Fixes #1237

### CODE REVIEW

File: src/Eval.hs [hunk 1]
```diff
@@ -397,11 +398,11 @@ eval ctx xobj@(XObj o info ty) preference resolver =
...
       case f' of
-        (ListPat l _) -> go l
-        (SymPat sym _) -> go sym
+        (ListPat l _) -> go l ResolveLocal
+        (SymPat sym _) -> go sym resolver
         _ -> pure (evalError ctx (format (GenericMalformed xobj)) (xobjInfo xobj))
-      where go x =
-              do (newCtx, f) <- eval ctx x preference ResolveLocal
+      where go x resolve =
+              do (newCtx, f) <- eval ctx x preference resolve
```
Comment (hellerve) [inline src/Eval.hs:405]: Are these changes part of the fix? They seem unrelated at a first glance

File: src/Eval.hs [hunk 2]
```diff
@@ -204,6 +204,7 @@ eval ctx xobj@(XObj o info ty) preference resolver =
             (WithPat (SymPat sym path) forms) -> specialCommandWith ctx sym path forms
             (DoPat forms) -> evaluateSideEffects forms
             (WhilePat cond body) -> specialCommandWhile ctx cond body
+            (SetPat iden value) -> specialCommandSet ctx (iden:[value])
```
Comment (hellerve) [inline src/Eval.hs:207]: Is this reordering an unrelated cleanup or important for this change?

File: src/Eval.hs (hunk 1 above)
Comment (scolsen) [reply src/Eval.hs:405]: Yes, they are. I didn't think it'd be relevant either, but if we use `ResolveLocal` on resolution of the `Sym` it results in hangs.

File: src/Eval.hs (hunk 2 above)
Comment (scolsen) [reply src/Eval.hs:207]: Yeah, this was just a small change for readability, I can undo it if we want a cleaner history.

Comment (hellerve) [reply src/Eval.hs:207]: No, that’s fine with me, I just wanted to check whether I was missing something!

**Pull request merged**